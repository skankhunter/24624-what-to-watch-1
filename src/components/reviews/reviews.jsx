import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {compose} from "redux";

import {Operation, actionClearReviews} from "../../reducer/reviews/reviews.js";

class Reviews extends PureComponent {
  constructor(props) {
    super(props);

    this._formDate = this._formDate.bind(this);
  }

  componentDidMount() {
    const {loadReviews, activeFilmId} = this.props;

    loadReviews(activeFilmId);
  }

  componentWillUnmount() {
    const {clearReviews} = this.props;

    clearReviews();
  }

  _formDate(dateStamp) {
    const date = new Date(dateStamp);
    const year = date.getFullYear();
    const month = date.getMonth();
    const day = date.getDate();

    return `${month} ${day}, ${year}.`;
  }

  render() {
    const {reviews} = this.props;

    return (
      <>
        <div className="movie-card__reviews movie-card__row">
          <div className="movie-card__reviews-col">
            {reviews.map((review, index) => {
              if (!(index % 2)) {
                return (
                  <div className="review" key={`${index}${review.user.name}`}>
                    <blockquote className="review__quote">
                      <p className="review__text">{review.comment}</p>

                      <footer className="review__details">
                        <cite className="review__author">
                          {review.user.name}
                        </cite>
                        <time className="review__date" dateTime={review.date}>
                          {this._formDate(review.date)}
                        </time>
                      </footer>
                    </blockquote>

                    <div className="review__rating">{review.rating}</div>
                  </div>
                );
              }

              return null;
            })}
          </div>
          <div className="movie-card__reviews-col">
            {reviews.map((review, index) => {
              if (index % 2) {
                return (
                  <div className="review" key={`${index}${review.user.name}`}>
                    <blockquote className="review__quote">
                      <p className="review__text">{review.comment}</p>

                      <footer className="review__details">
                        <cite className="review__author">
                          {review.user.name}
                        </cite>
                        <time className="review__date" dateTime={review.date}>
                          {this._formDate(review.date)}
                        </time>
                      </footer>
                    </blockquote>

                    <div className="review__rating">{review.rating}</div>
                  </div>
                );
              }
              return null;
            })}
          </div>
        </div>
      </>
    );
  }
}

Reviews.propTypes = {
  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        comment: PropTypes.string,
        date: PropTypes.string,
        id: PropTypes.number,
        rating: PropTypes.number,
        user: PropTypes.shape({
          name: PropTypes.string
        })
      })
  ).isRequired,
  loadReviews: PropTypes.func.isRequired,
  activeFilmId: PropTypes.number.isRequired,
  clearReviews: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  reviews: state.reviews.reviews
});
const mapDispatchToProps = (dispatch) => ({
  loadReviews: (filmId) => {
    dispatch(Operation.loadReviews(filmId));
  },
  clearReviews: () => {
    dispatch(actionClearReviews());
  }
});

export {Reviews};

export default compose(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )
)(Reviews);
