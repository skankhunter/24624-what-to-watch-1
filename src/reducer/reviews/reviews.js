const initialState = {
  reviews: [],
  reviewPostedStatus: false
};

const ActionType = {
  LOAD_REVIEWS: `LOAD_REVIEWS`,
  CLEAR_REVIEWS: `CLEAR_REVIEWS`,
  POST_REVIEW: `POST_REVIEW`
};

const actionLoadReviews = (loadedReviews) => {
  return {
    type: ActionType.LOAD_REVIEWS,
    payload: loadedReviews
  };
};

const actionClearReviews = () => {
  return {
    type: ActionType.CLEAR_REVIEWS
  };
};

const actionPostReview = (status) => {
  return {
    type: ActionType.POST_REVIEW,
    payload: status
  };
};

const operationLoadReviews = (filmId) => (dispatch, _getState, api) => {
  return api.get(`/comments/${filmId}`).then((response) => {
    dispatch(actionLoadReviews(response.data));
  });
};

const operationPostReview = (filmId, reviewInfo) => (dispatch, _getState, api) => {
  return api
    .post(`/comments/${filmId}`, reviewInfo)
    .then(() => {
      dispatch(actionPostReview(true));
    })
    .catch(() => {
      dispatch(actionPostReview(false));
    });
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_REVIEWS:
      return Object.assign({}, state, {
        reviews: action.payload.sort(function (reviewA, reviewB) {
          return new Date(reviewB.date) - new Date(reviewA.date);
        })
      });

    case ActionType.CLEAR_REVIEWS:
      return Object.assign({}, state, initialState);

    case ActionType.POST_REVIEW:
      return Object.assign({}, state, {
        reviewPostedStatus: action.payload
      });
  }

  return state;
};

export {
  ActionType,
  actionLoadReviews,
  actionClearReviews,
  actionPostReview,
  operationLoadReviews,
  operationPostReview,
  reducer
};
