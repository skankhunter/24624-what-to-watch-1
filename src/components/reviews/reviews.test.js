import React from "react";
import renderer from "react-test-renderer";

import {Reviews} from "./reviews.jsx";

const mocks = {
  activeFilmId: 3,
  onLoadReviews: jest.fn(),
  onClearReviews: jest.fn(),
  reviews: [
    {
      comment: `comment`,
      date: `20-12-2024`,
      id: 1,
      rating: 4,
      user: {
        name: `user`
      }
    },
    {
      comment: `comment2`,
      date: `20-12-2025`,
      id: 2,
      rating: 5,
      user: {
        name: `user2`
      }
    }
  ]
};

describe(`Reviews:`, () => {
  it(`Correctly renders after relaunch`, () => {
    const tree = renderer.create(<Reviews {...mocks} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
