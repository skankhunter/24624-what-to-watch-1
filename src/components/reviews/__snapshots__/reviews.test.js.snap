// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews: Correctly renders after relaunch 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          comment
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            user
          </cite>
          <time
            className="review__date"
            dateTime="20-12-2024"
          >
            NaN NaN, NaN.
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        4
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          comment2
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            user2
          </cite>
          <time
            className="review__date"
            dateTime="20-12-2025"
          >
            NaN NaN, NaN.
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        5
      </div>
    </div>
  </div>
</div>
`;
